File: src/main/java/com/example/year2016/Day01.java
package com.example.year2016;

import com.example.template.Day;
import com.example.utils.Pair;

import java.util.HashSet;
import java.util.Set;

public class Day01 extends Day {

    public Day01() {
        super("input.txt", "01", "2016");
    }

    static class Orientation {
        String direction = "N";
        int blocksX = 0;
        int blocksY = 0;
    }

    private void walk(Orientation o, char turnDirection, int steps) {
        switch (o.direction) {
            case "N" -> {
                o.direction = (turnDirection == 'R') ? "E" : "W";
                o.blocksX += (turnDirection == 'R') ? steps : -steps;
            }
            case "E" -> {
                o.direction = (turnDirection == 'R') ? "S" : "N";
                o.blocksY += (turnDirection == 'R') ? -steps : steps;
            }
            case "S" -> {
                o.direction = (turnDirection == 'R') ? "W" : "E";
                o.blocksX += (turnDirection == 'R') ? -steps : steps;
            }
            case "W" -> {
                o.direction = (turnDirection == 'R') ? "N" : "S";
                o.blocksY += (turnDirection == 'R') ? steps : -steps;
            }
        }
    }

    @Override
    public long part1()  {
        Orientation orientation = new Orientation();

        for (String s : getInput().get(0).trim().split(", ")) {
            char turnDirection = s.charAt(0);
            int steps = Integer.parseInt(s.substring(1));
            walk(orientation, turnDirection, steps);
        }

        return Math.abs(orientation.blocksX) + Math.abs(orientation.blocksY);
    }

    @Override
    public long part2() {
        Orientation orientation = new Orientation();
        Set<Pair<Integer, Integer>> visited = new HashSet<>();
        visited.add(new Pair<>(0, 0));
        int x = 0;
        int y = 0;

        for (String s : getInput().get(0).trim().split(", ")) {
            char turnDirection = s.charAt(0);
            int steps = Integer.parseInt(s.substring(1));
            walk(orientation, turnDirection, steps);

            for (int i = 0; i < steps; i++) {
                switch (orientation.direction) {
                    case "N" -> y++;
                    case "E" -> x++;
                    case "S" -> y--;
                    case "W" -> x--;
                }

                Pair<Integer, Integer> position = new Pair<>(x, y);
                if (visited.contains(position)) {
                    return Math.abs(x) + Math.abs(y);
                } else {
                    visited.add(position);
                }
            }
        }

        return 0;
    }
}

File: src/main/year2016/Day01.kt
package main.year2016

import java.io.File
import kotlin.math.abs

class Day01(text : String = "input.txt") {

    class Orientation{
        var direction = "N"
        var blocksX = 0
        var blocksY = 0
    }

    private val day = "01"
    private val year = "2016"

    private val input = File("src/resources/$year/$day/$text").readText()

    private fun walk(o: Orientation, turnDirection: Char, steps: Int) {
        when (o.direction) {
            "N" -> {
                o.direction = if (turnDirection == 'R') "E" else "W"
                o.blocksX += if (turnDirection == 'R') steps else -steps
            }
            "E" -> {
                o.direction = if (turnDirection == 'R') "S" else "N"
                o.blocksY += if (turnDirection == 'R') -steps else steps
            }
            "S" -> {
                o.direction = if (turnDirection == 'R') "W" else "E"
                o.blocksX += if (turnDirection == 'R') -steps else steps
            }
            "W" -> {
                o.direction = if (turnDirection == 'R') "N" else "S"
                o.blocksY += if (turnDirection == 'R') steps else -steps
            }
        }
    }

    fun part1() : Int = input.trim().split(", ").fold(Orientation()) { o, s -> walk(o, s[0], s.substring(1).toInt())
        o
    }.let { abs(it.blocksX) + abs(it.blocksY) }

    fun part2() : Int {
        var x = 0
        var y = 0
        return input.trim().split(", ").fold(Pair(Orientation(), mutableSetOf(Pair(0, 0)))) { (o, visited), s ->
            walk(o, s[0], s.substring(1).toInt())
            repeat(s.substring(1).toInt()) {
                when (o.direction) {
                    "N" -> y++
                    "E" -> x++
                    "S" -> y--
                    "W" -> x--
                }
                if (Pair(x, y) in visited) {
                    return abs(x) + abs(y)
                } else {
                    visited.add(Pair(x, y))
                }
            }
            Pair(o, visited)
        }.let { 0 }
    }
}

